File: backend/app/routers/game.py
from __future__ import annotations

from typing import Annotated, Any

from fastapi import APIRouter, Depends

from ..dependencies import Services, get_services
from ..schemas import (
    CreativeEventResolve,
    DailyChallengeSelect,
    GameCommandResponse,
    InventoryCommand,
    SkillIncrease,
    SpecializationSelect,
    WeeklyChallengeStart,
    WritingSessionStart,
)


router = APIRouter(prefix='/game', tags=['game'])


@router.get('/state', response_model=dict[str, Any])
def state(services: Annotated[Services, Depends(get_services)]):
    return services.game.get_state()


@router.get('/catalog', response_model=dict[str, Any])
def catalog(services: Annotated[Services, Depends(get_services)]):
    return services.game.get_shop_catalog()


@router.post('/writing-sessions/start', response_model=GameCommandResponse)
def start_writing_session(
        payload: WritingSessionStart,
        services: Annotated[Services, Depends(get_services)],
):
    return services.game.start_writing_session(
        payload.duration_minutes,
        payload.target_symbols,
        payload.intention,
        mode_key=payload.mode,
    )


@router.post('/writing-sessions/finish', response_model=GameCommandResponse)
def finish_writing_session(services: Annotated[Services, Depends(get_services)]):
    return services.game.finish_writing_session()


@router.post('/writing-sessions/cancel', response_model=GameCommandResponse)
def cancel_writing_session(services: Annotated[Services, Depends(get_services)]):
    return services.game.cancel_writing_session()


@router.post('/daily-challenge/select', response_model=GameCommandResponse)
def select_daily_challenge(
        payload: DailyChallengeSelect,
        services: Annotated[Services, Depends(get_services)],
):
    return services.game.select_daily_challenge(payload.option_id)


@router.post('/weekly-challenge/start', response_model=GameCommandResponse)
def start_weekly_challenge(
        payload: WeeklyChallengeStart,
        services: Annotated[Services, Depends(get_services)],
):
    return services.game.start_weekly_challenge(payload.challenge_id)


@router.post(
    '/inspiration-abilities/{ability_id}/activate',
    response_model=GameCommandResponse,
)
def activate_inspiration_ability(
        ability_id: str,
        services: Annotated[Services, Depends(get_services)],
):
    return services.game.activate_inspiration_ability(ability_id)


@router.post('/creative-events/resolve', response_model=GameCommandResponse)
def resolve_creative_event(
        payload: CreativeEventResolve,
        services: Annotated[Services, Depends(get_services)],
):
    return services.game.resolve_creative_event(payload.choice)


@router.post('/specialization/select', response_model=GameCommandResponse)
def select_specialization(
        payload: SpecializationSelect,
        services: Annotated[Services, Depends(get_services)],
):
    return services.game.select_specialization(payload.specialization_id)


@router.post('/specialization/ability/activate', response_model=GameCommandResponse)
def activate_specialization_ability(
        services: Annotated[Services, Depends(get_services)],
):
    return services.game.activate_specialization_ability()


@router.post('/skills/{skill_id}/increase', response_model=GameCommandResponse)
def increase_skill(
        skill_id: str,
        payload: SkillIncrease,
        services: Annotated[Services, Depends(get_services)],
):
    return services.game.increase_skill(skill_id, payload.points)


@router.post('/quests/{quest_id}/start', response_model=GameCommandResponse)
def start_quest(
        quest_id: str,
        services: Annotated[Services, Depends(get_services)],
):
    return services.game.start_quest(quest_id)


@router.post('/quests/{quest_id}/abandon', response_model=GameCommandResponse)
def abandon_quest(
        quest_id: str,
        services: Annotated[Services, Depends(get_services)],
):
    return services.game.abandon_quest(quest_id)


@router.post('/inventory/buy', response_model=GameCommandResponse)
def buy_item(
        payload: InventoryCommand,
        services: Annotated[Services, Depends(get_services)],
):
    return services.game.buy_item(payload.category, payload.item_id, payload.count)


@router.post('/inventory/sell', response_model=GameCommandResponse)
def sell_item(
        payload: InventoryCommand,
        services: Annotated[Services, Depends(get_services)],
):
    return services.game.sell_item(payload.category, payload.item_id, payload.count)


@router.post('/inventory/use', response_model=GameCommandResponse)
def use_item(
        payload: InventoryCommand,
        services: Annotated[Services, Depends(get_services)],
):
    return services.game.use_item(payload.category, payload.item_id, payload.count)

File: frontend/src/api/game.ts
import { apiRequest } from './client'
import { invoke } from '@tauri-apps/api/core'
import { currentPlatform } from '@/platform/runtime'
import type {
  BankProductRequest,
  GameCommandResponse,
  DeveloperModeState,
  DeveloperProfileUpdate,
  GameNotifications,
  GameState,
  InventoryCommand,
  ShopCatalog,
  WritingSessionStart,
} from '@/types/game'

const GAME_PATH = '/api/game'

const nativeGame = <T>(commandName: string, payload?: Record<string, unknown>): Promise<T> =>
  invoke<T>(commandName, payload === undefined ? undefined : payload)

function isDesktopGame(): boolean {
  return currentPlatform() === 'tauri'
}

function command(path: string, body?: unknown, method = 'POST'): Promise<GameCommandResponse> {
  return apiRequest<GameCommandResponse>(`${GAME_PATH}${path}`, { method, body })
}

function awardPath(awardId: string): string {
  return `/custom-awards/${encodeURIComponent(awardId)}`
}

export const gameApi = {
  state(signal?: AbortSignal): Promise<GameState> {
    if (isDesktopGame()) return nativeGame<GameState>('game_state')
    return apiRequest<GameState>(`${GAME_PATH}/state`, { signal })
  },

  notifications(signal?: AbortSignal): Promise<GameNotifications> {
    if (isDesktopGame()) return nativeGame<GameNotifications>('game_notifications')
    return apiRequest<GameNotifications>(`${GAME_PATH}/notifications`, { signal })
  },

  markNotificationRead(notificationId: string): Promise<GameNotifications> {
    if (isDesktopGame()) return nativeGame<GameNotifications>('mark_game_notification_read', { notificationId })
    return apiRequest<GameNotifications>(
      `${GAME_PATH}/notifications/${encodeURIComponent(notificationId)}/read`,
      { method: 'POST' },
    )
  },

  markAllNotificationsRead(): Promise<GameNotifications> {
    if (isDesktopGame()) return nativeGame<GameNotifications>('mark_all_game_notifications_read')
    return apiRequest<GameNotifications>(`${GAME_PATH}/notifications/read-all`, {
      method: 'POST',
    })
  },

  catalog(signal?: AbortSignal): Promise<ShopCatalog> {
    if (isDesktopGame()) return nativeGame<ShopCatalog>('game_catalog')
    return apiRequest<ShopCatalog>(`${GAME_PATH}/catalog`, { signal })
  },

  developerState(): Promise<DeveloperModeState> {
    if (isDesktopGame()) return nativeGame<DeveloperModeState>('game_developer_state')
    return apiRequest<DeveloperModeState>(`${GAME_PATH}/developer`)
  },

  updateDeveloperProfile(payload: DeveloperProfileUpdate): Promise<GameCommandResponse> {
    if (isDesktopGame()) return nativeCommand('game_update_developer_profile', {
      level: payload.level, health: payload.health, coins: payload.coins, exp: payload.exp,
      testDateEnabled: payload.test_date_enabled, testDatetime: payload.test_datetime,
    })
    return command('/developer/profile', payload, 'PUT')
  },

  grantDeveloperInventoryItem(
    category: string,
    itemId: string,
    count: number,
  ): Promise<GameCommandResponse> {
    if (isDesktopGame()) return nativeCommand('game_grant_developer_inventory_item', {
      category, itemId, count,
    })
    return command('/developer/inventory', {
      category,
      item_id: itemId,
      count,
    })
  },

  startWritingSession(payload: WritingSessionStart): Promise<GameCommandResponse> {
    if (isDesktopGame()) return nativeCommand('game_start_writing_session', {
      durationMinutes: payload.duration_minutes,
      targetSymbols: payload.target_symbols,
      intention: payload.intention,
      mode: payload.mode,
    })
    return command('/writing-sessions/start', payload)
  },

  finishWritingSession(): Promise<GameCommandResponse> {
    if (isDesktopGame()) return nativeCommand('game_finish_writing_session')
    return command('/writing-sessions/finish')
  },

  cancelWritingSession(): Promise<GameCommandResponse> {
    if (isDesktopGame()) return nativeCommand('game_cancel_writing_session')
    return command('/writing-sessions/cancel')
  },

  applyStreakFreeze(
    target: 'global' | 'project',
    projectId?: string,
  ): Promise<GameCommandResponse> {
    if (isDesktopGame()) return nativeCommand('game_apply_streak_freeze', { target, projectId })
    return command('/streak-freezes/apply', {
      target,
      ...(projectId ? { project_id: projectId } : {}),
    })
  },

  selectDailyChallenge(optionId: string): Promise<GameCommandResponse> {
    if (isDesktopGame()) return nativeCommand('game_select_daily_challenge', { optionId })
    return command('/daily-challenge/select', { option_id: optionId })
  },

  startWeeklyChallenge(challengeId: string): Promise<GameCommandResponse> {
    if (isDesktopGame()) return nativeCommand('game_start_weekly_challenge', { challengeId })
    return command('/weekly-challenge/start', { challenge_id: challengeId })
  },

  activateInspirationAbility(abilityId: string): Promise<GameCommandResponse> {
    if (isDesktopGame()) return nativeCommand('game_activate_inspiration_ability', { abilityId })
    return command(`/inspiration-abilities/${encodeURIComponent(abilityId)}/activate`)
  },

  resolveCreativeEvent(choice: 'safe' | 'risk'): Promise<GameCommandResponse> {
    if (isDesktopGame()) return nativeCommand('game_resolve_creative_event', { choice })
    return command('/creative-events/resolve', { choice })
  },

  selectSpecialization(specializationId: string): Promise<GameCommandResponse> {
    if (isDesktopGame()) return nativeCommand('game_select_specialization', { specializationId })
    return command('/specialization/select', { specialization_id: specializationId })
  },

  activateSpecializationAbility(): Promise<GameCommandResponse> {
    if (isDesktopGame()) return nativeCommand('game_activate_specialization_ability')
    return command('/specialization/ability/activate')
  },

  increaseSkill(skillId: string, points = 1): Promise<GameCommandResponse> {
    if (isDesktopGame()) return nativeCommand('game_increase_skill', { skillId, points })
    return command(`/skills/${encodeURIComponent(skillId)}/increase`, { points })
  },

  startQuest(questId: string): Promise<GameCommandResponse> {
    if (isDesktopGame()) return nativeCommand('game_start_quest', { questId })
    return command(`/quests/${encodeURIComponent(questId)}/start`)
  },

  abandonQuest(questId: string): Promise<GameCommandResponse> {
    if (isDesktopGame()) return nativeCommand('game_abandon_quest', { questId })
    return command(`/quests/${encodeURIComponent(questId)}/abandon`)
  },

  buyItem(payload: InventoryCommand): Promise<GameCommandResponse> {
    if (isDesktopGame()) return nativeCommand('game_buy_item', nativeInventoryPayload(payload))
    return command('/inventory/buy', payload)
  },

  sellItem(payload: InventoryCommand): Promise<GameCommandResponse> {
    if (isDesktopGame()) return nativeCommand('game_sell_item', nativeInventoryPayload(payload))
    return command('/inventory/sell', payload)
  },

  useItem(payload: InventoryCommand): Promise<GameCommandResponse> {
    if (isDesktopGame()) return nativeCommand('game_use_item', nativeInventoryPayload(payload))
    return command('/inventory/use', payload)
  },

  createCustomAward(name: string, price: number): Promise<GameCommandResponse> {
    if (isDesktopGame()) return nativeCommand('game_create_custom_award', { name, price })
    return command('/custom-awards', { name, price })
  },

  updateCustomAward(
    awardId: string,
    payload: { name?: string; price?: number },
  ): Promise<GameCommandResponse> {
    if (isDesktopGame()) return nativeCommand('game_update_custom_award', { awardId, ...payload })
    return command(awardPath(awardId), payload, 'PATCH')
  },

  deleteCustomAward(awardId: string): Promise<GameCommandResponse> {
    if (isDesktopGame()) return nativeCommand('game_delete_custom_award', { awardId })
    return command(awardPath(awardId), undefined, 'DELETE')
  },

  buyCustomAward(awardId: string, count = 1): Promise<GameCommandResponse> {
    if (isDesktopGame()) return nativeCommand('game_buy_custom_award', { awardId, count })
    return command(`${awardPath(awardId)}/buy`, { count })
  },

  sellCustomAward(awardId: string, count = 1): Promise<GameCommandResponse> {
    if (isDesktopGame()) return nativeCommand('game_sell_custom_award', { awardId, count })
    return command(`${awardPath(awardId)}/sell`, { count })
  },

  useCustomAward(awardId: string, count = 1): Promise<GameCommandResponse> {
    if (isDesktopGame()) return nativeCommand('game_use_custom_award', { awardId, count })
    return command(`${awardPath(awardId)}/use`, { count })
  },

  previewBankProduct(payload: BankProductRequest): Promise<GameCommandResponse> {
    if (isDesktopGame()) return nativeCommand('game_preview_bank_product', {
      productType: payload.product_type,
      amount: payload.amount,
      days: payload.days,
      allowInterestWithdrawal: payload.allow_interest_withdrawal,
    })
    return command('/bank/preview', payload)
  },

  openBankCredit(amount: number, days: number): Promise<GameCommandResponse> {
    if (isDesktopGame()) return nativeCommand('game_open_bank_credit', { amount, days })
    return command('/bank/credit', { amount, days })
  },

  openBankDeposit(
    amount: number,
    days: number,
    allowInterestWithdrawal: boolean,
  ): Promise<GameCommandResponse> {
    if (isDesktopGame()) return nativeCommand('game_open_bank_deposit', {
      amount, days, allowInterestWithdrawal,
    })
    return command('/bank/deposit', {
      amount,
      days,
      allow_interest_withdrawal: allowInterestWithdrawal,
    })
  },

  processBankEvents(autoPay = true): Promise<GameCommandResponse> {
    if (isDesktopGame()) return nativeGame<GameCommandResponse>('game_process_bank_events', { autoPay })
    return command('/bank/process', { auto_pay: autoPay })
  },

  makeBankLoanPayment(): Promise<GameCommandResponse> {
    if (isDesktopGame()) return nativeCommand('game_make_bank_loan_payment')
    return command('/bank/credit/payment')
  },

  partiallyRepayBankCredit(amount: number): Promise<GameCommandResponse> {
    if (isDesktopGame()) return nativeCommand('game_partially_repay_bank_credit', { amount })
    return command('/bank/credit/partial-repayment', { amount })
  },

  repayBankCredit(): Promise<GameCommandResponse> {
    if (isDesktopGame()) return nativeCommand('game_repay_bank_credit')
    return command('/bank/credit/repay')
  },

  topUpBankDeposit(amount: number): Promise<GameCommandResponse> {
    if (isDesktopGame()) return nativeCommand('game_top_up_bank_deposit', { amount })
    return command('/bank/deposit/top-up', { amount })
  },

  withdrawBankDeposit(allowEarly = false): Promise<GameCommandResponse> {
    if (isDesktopGame()) return nativeCommand('game_withdraw_bank_deposit', { allowEarly })
    return command('/bank/deposit/withdraw', { allow_early: allowEarly })
  },

  withdrawBankDepositInterest(): Promise<GameCommandResponse> {
    if (isDesktopGame()) return nativeCommand('game_withdraw_bank_interest')
    return command('/bank/deposit/interest/withdraw')
  },
}

function nativeCommand(
  commandName: string,
  payload?: Record<string, unknown>,
): Promise<GameCommandResponse> {
  return nativeGame<GameCommandResponse>(commandName, payload)
}

function nativeInventoryPayload(payload: InventoryCommand): Record<string, unknown> {
  return { category: payload.category, itemId: payload.item_id, count: payload.count }
}
